Userland/Libraries/LibC/pwd.cpp

```diff
@@ -134,123 +128,57 @@ int getpwent_r(struct passwd* passwd_buf, char* buffer, size_t buffer_size, stru
 
     while (true) {
         if (!s_stream || feof(s_stream)) {
-            errno = EIO;
-            return -1;
+            *passwd_entry_ptr = nullptr;
+            return ENOENT;
         }
 
         if (ferror(s_stream)) {
-            dbgln("getpwent(): Read error: {}", strerror(ferror(s_stream)));
-            errno = EIO;
-            return -1;
+            *passwd_entry_ptr = nullptr;
+            return ferror(s_stream);
         }
 
         ++s_line_number;
         char* s = fgets(buffer, buffer_size, s_stream);
 
-        // Silently tolerate an empty line at the end.
         if ((!s || !s[0]) && feof(s_stream)) {
             *passwd_entry_ptr = nullptr;
-            return 0;
+            return ENOENT;
         }
 
         if (strlen(s) == buffer_size - 1) {
-            errno = ERANGE;
-            return -1;
+            *passwd_entry_ptr = nullptr;
+            return ERANGE;
         }
 
         if (parse_pwddb_entry(buffer, *passwd_buf)) {
             *passwd_entry_ptr = passwd_buf;
             return 0;
         }
-        // Otherwise, proceed to the next line.
     }
 }
 
-static void construct_pwd(struct passwd* pwd, char* buf, struct passwd** result)
+// https://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwnam.html
+int getpwnam_r(char const* name, struct passwd* pwd, char* buffer, size_t bufsize, struct passwd** result)
 {
-    auto* buf_name = &buf[0];
-    auto* buf_passwd = &buf[s_name.length() + 1];
-    auto* buf_gecos = &buf[s_name.length() + 1 + s_gecos.length() + 1];
-    auto* buf_dir = &buf[s_gecos.length() + 1 + s_name.length() + 1 + s_gecos.length() + 1];
-    auto* buf_shell = &buf[s_dir.length() + 1 + s_gecos.length() + 1 + s_name.length() + 1 + s_gecos.length() + 1];
-
-    bool ok = true;
-    ok = ok && s_name.copy_characters_to_buffer(buf_name, s_name.length() + 1);
-    ok = ok && s_passwd.copy_characters_to_buffer(buf_passwd, s_passwd.length() + 1);
-    ok = ok && s_gecos.copy_characters_to_buffer(buf_gecos, s_gecos.length() + 1);
-    ok = ok && s_dir.copy_characters_to_buffer(buf_dir, s_dir.length() + 1);
-    ok = ok && s_shell.copy_characters_to_buffer(buf_shell, s_shell.length() + 1);
-
-    VERIFY(ok);
-
-    *result = pwd;
-    pwd->pw_name = buf_name;
-    pwd->pw_passwd = buf_passwd;
-    pwd->pw_gecos = buf_gecos;
-    pwd->pw_dir = buf_dir;
-    pwd->pw_shell = buf_shell;
-}
-
-int getpwnam_r(char const* name, struct passwd* pwd, char* buf, size_t buflen, struct passwd** result)
-{
-    // FIXME: This is a HACK!
-    TemporaryChange name_change { s_name, {} };
-    TemporaryChange passwd_change { s_passwd, {} };
-    TemporaryChange gecos_change { s_gecos, {} };
-    TemporaryChange dir_change { s_dir, {} };
-    TemporaryChange shell_change { s_shell, {} };
-
     setpwent();
```

Comment:
Is setpwent still needed? I can see that getpwent_r does need it, but calls it by itself.

Reply:
`getpwent_r()` only calls it if the database is not open at all.

We must call it here to ensure that the database is rewound to the start, otherwise we may not see all entries.